### PR TITLE
feature: support take over CRI meta when d2p migration

### PR DESCRIPTION
**Pull request by HusterWan**
Signed-off-by: ziren.wzr <ziren.wzr@alibaba-inc.com>

Since we have scenes need to upgrade docker to PouchContainer when the CRI is on, so we should support to take over the CRI meta when migration.

This PR contains two parts:
  * The first commit is to support generate CRI meta by container meta.
  * The second commit is to take over CRI sandbox info when migration.

### CODE REVIEW

File: migrator/cri_migrate.go [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	// prepare sandbox dir if not exist.
+	if _, err := os.Stat(sandboxDir); err != nil && os.IsNotExist(err) {
+		if err := os.MkdirAll(sandboxDir, 0666); err != nil {
+			return fmt.Errorf("failed to prepare sandbox dir %s: %v", sandboxDir, err)
+		}
+	}
+
+	dnsConfig := sandbox.Config.DnsConfig
+	if dnsConfig != nil {
+		resolvContent, err = convertor.GenerateResolvConfigContent(dnsConfig)
```
Comment (zhuangqh) [inline migrator/cri_migrate.go]: Just copy the resolve file directly. It is unnecessary to regenerate.

File: pouch/convertor/cri_test.go [hunk 2]
```diff
@@ -0,0 +1,309 @@
...
+		t.Errorf("got wrong PodSandboxConfig, want %v, got %v", wantCRIMeta.Config, gotCRIMeta.Config)
+	}
+}
+
+func Test_toDNSConfig(t *testing.T) {
+	var (
+		resolvContent = `search hz.ali.com
+nameserver 10.65.1.3
+nameserver  30.14.130.2
+nameserver 10.65.1.1
```
Comment (zhuangqh) [inline pouch/convertor/cri_test.go]: I am afraid that there would be some security issue. just using fake ip in test.

File: pouch/convertor/cri_test.go [hunk 3]
```diff
@@ -0,0 +1,309 @@
...
+				name: "k8s_POD_mysql_default_xx888&&&_0",
+			},
+			want: &runtime.PodSandboxMetadata{
+				Name:      "mysql",
+				Namespace: "default",
+				Uid:       "xx888&&&",
+				Attempt:   uint32(0),
+			},
+			wantErr: false,
+		},
```
Comment (zhuangqh) [inline pouch/convertor/cri_test.go]: missing case: sandbox with a non int attempt
`k8s_POD_mysql_default_98e2377b-f2ea-11e8-89af-6c92bf2cf061_nonint`

File: pouch/convertor/cri_test.go [hunk 4]
```diff
@@ -0,0 +1,309 @@
...
+				return
+			}
+			if !reflect.DeepEqual(got, tt.want) {
+				t.Errorf("getPodSandboxMetadataBySandboxName() = %v, want %v", got, tt.want)
+			}
+		})
+	}
+}
+
+func TestGenerateResolvConfigContent(t *testing.T) {
```
Comment (zhuangqh) [inline pouch/convertor/cri_test.go]: unnecessary

File: utils/fileutils.go [hunk 5]
```diff
@@ -34,3 +34,21 @@ func IsDirEmpty(dir string) (bool, error) {
 	}
 	return false, nil
 }
+
+// CopyFile copys src file to dest file
+func CopyFile(src, dst string, perm os.FileMode) error {
```
Comment (zhuangqh) [inline utils/fileutils.go]: In UNIX world, we usually put dst in front of src 😅

File: utils/utils.go [hunk 6]
```diff
@@ -54,3 +55,45 @@ func StringInSlice(input []string, str string) bool {
...
+		if str != "" {
+			output = append(output, str)
+		}
+	}
+
+	return output
+}
+
+// StringSliceEqual compare two string slice, ignore the order.
+func StringSliceEqual(s1, s2 []string) bool {
```
Comment (zhuangqh) [inline utils/utils.go]: `[]string{"a", "a", "b"}` is not equal to `[]string{"a", "b","b"}`. but `StringSliceEqual` return true

File: utils/fileutils.go (hunk 5 above)
Comment (HusterWan) [reply utils/fileutils.go]: 👍 😆 Good advice, updated

File: utils/utils.go (hunk 6 above)
Comment (HusterWan) [reply utils/utils.go]: Great catch

File: pouch/convertor/cri.go [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+
+const (
+	// PouchContainerTypeLabelKey label to identify whether a container is a sandbox
+	// or container
+	PouchContainerTypeLabelKey = "io.kubernetes.pouch.type"
+	// PouchContainerTypeLabelSandbox specify this container is a sandbox
+	PouchContainerTypeLabelSandbox = "sandbox"
+	// PouchContainerTypeLabelContainer specify this container is a container
+	PouchContainerTypeLabelContainer = "container"
+
```
Comment (Starnop) [inline pouch/convertor/cri.go]: We will add a label in the future: containerLogPathLabelKey = "io.kubernetes.container.logpath"

File: pouch/convertor/cri.go [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+}
+
+func toLinuxSandboxSecurityContext(c *localtypes.Container) (*runtime.LinuxSandboxSecurityContext, error) {
+	linuxSecurityContext := &runtime.LinuxSandboxSecurityContext{
+		NamespaceOptions: &runtime.NamespaceOption{
+			Pid: runtime.NamespaceMode_CONTAINER,
+		},
+	}
+	if c.NetworkSettings != nil {
+		if _, ok := c.NetworkSettings.Networks["host"]; ok {
```
Comment (Starnop) [inline pouch/convertor/cri.go]: Is it appropriate to set all to host mode according to the NetworkSettings？

File: pouch/convertor/cri.go [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+	}
+
+	// ID
+	sandbox := &localtypes.SandboxMeta{
+		ID: c.ID,
+	}
+
+	// Runtime and LxcfsEnabled
+	if c.HostConfig != nil {
+		sandbox.Runtime = c.HostConfig.Runtime
```
[the remark this answers is not in the document]
Comment (HusterWan) [reply pouch/convertor/cri.go]: We should consider how to keep compatibility with the old pods

File: pouch/convertor/cri.go (hunk 8 above)
Comment (HusterWan) [reply pouch/convertor/cri.go]: actually, we indeed set all to host mode if the pod is a daemonset.

**Comment by HusterWan** [comment]
Since we have already verified this feature in the production environment. So  I will merge this pr.

**Pull request merged**